extensions-core/lookups-cached-single/src/main/java/io/druid/server/lookup/jdbc/JdbcDataFetcher.java

```diff
@@ -0,0 +1,230 @@
+  private static final int DEFAULT_STREAMING_FETCH_SIZE = 1000;
+
+  @JsonProperty
+  private final MetadataStorageConnectorConfig connectorConfig;
+  @JsonProperty
+  private final String table;
+  @JsonProperty
+  private final String keyColumn;
+  @JsonProperty
+  private final String valueColumn;
+  @JsonProperty
+  private final int streamingFetchSize;
+
+  private final String fetchAllQuery;
+  private final String fetchQuery;
+  private final String reverseFetchQuery;
+  private final DBI dbi;
+
+  public JdbcDataFetcher(
+      @JsonProperty("connectorConfig") MetadataStorageConnectorConfig connectorConfig,
+      @JsonProperty("table") String table,
+      @JsonProperty("keyColumn") String keyColumn,
+      @JsonProperty("valueColumn") String valueColumn,
+      @JsonProperty("streamingFetchSize") Integer streamingFetchSize
+  )
+  {
+    this.connectorConfig = Preconditions.checkNotNull(connectorConfig, "connectorConfig");
+    this.streamingFetchSize = streamingFetchSize == null ? DEFAULT_STREAMING_FETCH_SIZE : streamingFetchSize;
+    Preconditions.checkNotNull(connectorConfig.getConnectURI(), "connectorConfig.connectURI");
+    this.table = Preconditions.checkNotNull(table, "table");
+    this.keyColumn = Preconditions.checkNotNull(keyColumn, "keyColumn");
+    this.valueColumn = Preconditions.checkNotNull(valueColumn, "valueColumn");
+
+    this.fetchAllQuery = String.format(
+        "SELECT %s, %s FROM %s",
+        this.keyColumn,
+        this.valueColumn,
+        this.table
+    );
+    this.fetchQuery = String.format(
+        "SELECT %s FROM %s WHERE %s = :val",
+        this.valueColumn,
+        this.table,
+        this.keyColumn
+    );
+    this.reverseFetchQuery = String.format(
+        "SELECT %s FROM %s WHERE %s = :val",
+        this.keyColumn,
+        this.table,
+        this.valueColumn
+    );
+    dbi = new DBI(
+        connectorConfig.getConnectURI(),
+        connectorConfig.getUser(),
+        connectorConfig.getPassword()
+    );
+    dbi.registerMapper(new KeyValueResultSetMapper(keyColumn, valueColumn));
+  }
+
+  @Override
+  public Iterable<Map.Entry<String, String>> fetchAll()
+  {
+    return inReadOnlyTransaction(new TransactionCallback<List<Map.Entry<String, String>>>()
+                                 {
+                                   @Override
+                                   public List<Map.Entry<String, String>> inTransaction(
+                                       Handle handle,
+                                       TransactionStatus status
+                                   ) throws Exception
+                                   {
+                                     return handle.createQuery(fetchAllQuery)
+                                                  .setFetchSize(streamingFetchSize)
+                                                  .map(new KeyValueResultSetMapper(keyColumn, valueColumn))
+                                                  .list();
```

Comment:
It would be cool if the results could stream into the cache, but that's probably beyond the scope of this PR.